### PR TITLE
chore: Make IPFS instances DRY way

### PR DESCRIPTION
**Pull request by ukstv**
This started as "I want to create an IPFS instance in my test for LocalPinApi".

- Extract `createIPFS`: creates an IPFS instance with random repository folder, and random port,
- Extract `swarmConnect`: connects two IPFS instances together via `swarm.connect`, eliminates need for `multiaddr` variables in tests.

FYI Found problematic test wrt #754, commented it for now, made a note in #754.

### CODE REVIEW

File: packages/core/src/__tests__/ceramic-anchor.test.ts [hunk 1]
```diff
@@ -2,35 +2,14 @@ import Ceramic from '../ceramic'
...
 import tmp from 'tmp-promise'
-import IPFS from 'ipfs-core'
 import * as u8a from 'uint8arrays'
-
-import getPort from 'get-port'
-
-import dagJose from 'dag-jose'
-import basicsImport from 'multiformats/cjs/src/basics-import.js'
-import legacy from 'multiformats/cjs/src/legacy.js'
+import * as _ from 'lodash'
```
Comment (stbrody) [inline packages/core/src/__tests__/ceramic-anchor.test.ts:6]: this seems potentially confusing.  Is this a common convention in javascript code to import a package as `_`?  Wouldn't it be clearer just to call it `lodash`?

File: packages/core/src/__tests__/ceramic-api.test.ts [hunk 2]
```diff
@@ -584,20 +557,21 @@ describe('Ceramic API', () => {
...
-        },
-        {
-          docId: notExistDocId,
-          paths: ['/e/f' , '/d']
-        }
-      ]
-      const docs = await ceramic.multiQuery(queries, 1000)
-      expect(Object.keys(docs).length).toEqual(3)
-    })
+    // FIXME 754 Problematic test for #754
```
Comment (stbrody) [inline packages/core/src/__tests__/ceramic-api.test.ts:560]: How did the change from this PR make this test worse?  It looks totally unrelated to me?

File: packages/core/src/__tests__/ceramic-api.test.ts [hunk 3]
```diff
@@ -467,6 +439,7 @@ describe('Ceramic API', () => {
     const notExistDocId = DocID.fromString('kjzl6cwe1jw1495fyn7770ujykvl1f8sskbzsevlux062ajragz9hp3akdqbmdg')
 
     beforeAll(async () => {
+      ceramic = await createCeramic()
```
Comment (stbrody) [inline packages/core/src/__tests__/ceramic-api.test.ts:442]: ugh.  So I'm guessing these tests worked if they were run after the previous tests, but failed if run on their own because the `ceramic` object hadn't been set?  In any case if we're adding this to `beforeAll` we probably need a corresponding `afterAll` to close it right?

File: packages/core/src/__tests__/ceramic-api.test.ts [hunk 4]
```diff
@@ -96,20 +76,12 @@ describe('Ceramic API', () => {
...
-        Bootstrap: []
-      }
-    })
+    ipfs = await createIPFS()
   })
 
   afterAll(async () => {
     await ipfs.stop(() => console.log('IPFS stopped'))
-    await tmpFolder.cleanup()
+    await ceramic?.close()
```
Comment (stbrody) [inline packages/core/src/__tests__/ceramic-api.test.ts:84]: Should we change this from an `afterAll` to an `afterEach`?  Otherwise if a test reassigns the `ceramic` variable without closing the existing one then the previous ceramic gets leaked

File: packages/core/src/__tests__/ipfs-util.ts [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+  };
+
+  const config = _.merge({}, defaultConfig, overrideConfig);
+  const instance = await IPFS.create(config);
+
+  // IPFS does not notify you when it stops.
+  // Here we intercept a call to `ipfs.stop` to clean up IPFS repository folder.
+  // Poor man's hook.
+  return new Proxy(instance, {
+    get(target: any, p: PropertyKey): any {
```
Comment (stbrody) [inline packages/core/src/__tests__/ipfs-util.ts:36]: woah.  javascript really is a wild language

**Review by stbrody — COMMENTED** [review]
Great cleanup, thanks for doing it!  A couple of small questions inline but nothing big

File: packages/core/src/__tests__/ceramic-anchor.test.ts (hunk 1 above)
Comment (ukstv) [reply packages/core/src/__tests__/ceramic-anchor.test.ts:6]: It is a convention. Comes from times of a package called [underscore](https://underscorejs.org). Lodash is effectively a replacement for underscore, that's why "_".

File: packages/core/src/__tests__/ceramic-api.test.ts (hunk 2 above)
Comment (ukstv) [reply packages/core/src/__tests__/ceramic-api.test.ts:560]: Totally unrelated feature-wise. At one point I anticipated dropping "--runInBand", and this test still consumer resources after "ceramic.close".

For future reference: multi-query timeout is a shallow one. The timeout does not propagate to `Document.load`, so it hangs till IPFS timeout expires.

Comment (ukstv) [reply packages/core/src/__tests__/ceramic-api.test.ts:560]: Going to put it back in a few.

Comment (ukstv) [reply packages/core/src/__tests__/ceramic-api.test.ts:560]: CC @zachferland about multi-query timeout.

File: packages/core/src/__tests__/ceramic-api.test.ts (hunk 3 above)
Comment (ukstv) [reply packages/core/src/__tests__/ceramic-api.test.ts:442]: Yes! Sad thing about it I has just wiped the changes that make it concur with your logic. Restoring.

File: packages/core/src/__tests__/ceramic-api.test.ts (hunk 4 above)
Comment (ukstv) [reply packages/core/src/__tests__/ceramic-api.test.ts]: Changed.

File: packages/core/src/__tests__/ceramic-api.test.ts (hunk 3 above)
Comment (ukstv) [reply packages/core/src/__tests__/ceramic-api.test.ts:413]: And, restored.

File: packages/core/src/__tests__/ceramic-api.test.ts (hunk 2 above)
Comment (ukstv) [reply packages/core/src/__tests__/ceramic-api.test.ts]: Uncommented back.

File: packages/core/src/__tests__/ceramic-anchor.test.ts [hunk 6]
```diff
@@ -72,55 +51,21 @@ describe('Ceramic anchoring', () => {
...
-      return await getPort({port: getPort.makeRange(start + 1, start + offset)})
-    }
-
-    ([port1, port2, port3] = await Promise.all([p1Start, p2Start, p3Start].map(start => findPort(start, pOffset))));
-    ([ipfs1, ipfs2, ipfs3] = await Promise.all([port1, port2, port3].map(port => createIPFS(buildConfig(tmpFolder.path, port)))));
-    ([p1Start, p2Start, p3Start] = [p1Start, p2Start, p3Start].map(start => start + pOffset))
-
-    multaddr1 = (await ipfs1.id()).addresses[0].toString()
-    multaddr2 = (await ipfs2.id()).addresses[0].toString()
+    [ipfs1, ipfs2, ipfs3] = await Promise.all(_.times(3).map(() => createIPFS()));
```
Comment (carlbarrdahl) [inline packages/core/src/__tests__/ceramic-anchor.test.ts:58]: You can use this instead of the lodash dep: `Array.from({length: 3}).map()`

File: packages/core/src/__tests__/ceramic-anchor.test.ts (hunk 1 above)
Comment (stbrody) [reply packages/core/src/__tests__/ceramic-anchor.test.ts:6]: Ah okay.  Yeah I guess this package provides such fundamental utilities it's reasonable to give it someone special treatment.  Sounds good

**Review by stbrody — APPROVED** [review]
LGTM

File: packages/core/src/__tests__/ceramic-anchor.test.ts (hunk 1 above)
Comment (ukstv) [reply packages/core/src/__tests__/ceramic-anchor.test.ts:6]: At the time it was groundbreaking indeed. Removing it away.

Comment (ukstv) [reply packages/core/src/__tests__/ceramic-anchor.test.ts:6]: Was that sarcasm in "fundamental"?

Comment (stbrody) [reply packages/core/src/__tests__/ceramic-anchor.test.ts:6]: No, it wasn't sarcastic, I was being genuine.  It seems like this library is designed to give some pretty basic core primitives that are generally useful and can almost be thought of as extensions to the language's standard library.
